File: core/src/com/qtech/bubbleblaster/LoadedGame.java
package com.qtech.bubbleblaster;

import com.qtech.bubbleblaster.common.GraphicsProcessor;
import com.qtech.bubbleblaster.common.entity.*;
import com.qtech.bubbleblaster.common.gamestate.GameEvent;
import com.qtech.bubbleblaster.common.gametype.AbstractGameType;
import com.qtech.bubbleblaster.common.random.PseudoRandom;
import com.qtech.bubbleblaster.common.screen.Screen;
import com.qtech.bubbleblaster.core.common.SavedGame;
import com.qtech.bubbleblaster.core.controllers.KeyboardController;
import com.qtech.bubbleblaster.core.utils.categories.GraphicsUtils;
import com.qtech.bubbleblaster.core.utils.categories.ShapeUtils;
import com.qtech.bubbleblaster.environment.Environment;
import com.qtech.bubbleblaster.event.CollisionEvent;
import com.qtech.bubbleblaster.event.KeyboardEvent;
import com.qtech.bubbleblaster.event.MouseEvent;
import com.qtech.bubbleblaster.event.SubscribeEvent;
import com.qtech.bubbleblaster.event.bus.EventBus;
import com.qtech.bubbleblaster.event.type.KeyEventType;
import com.qtech.bubbleblaster.event.type.MouseEventType;
import com.qtech.bubbleblaster.media.AudioSlot;
import com.qtech.bubbleblaster.screen.CommandScreen;
import com.qtech.bubbleblaster.screen.GameOverScreen;
import com.qtech.bubbleblaster.screen.PauseScreen;
import com.qtech.bubbleblaster.util.Util;
import lombok.Getter;

import java.io.File;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class LoadedGame implements Runnable {

    // Fonts.
    private final Font defaultFont = new Font(Util.getGame().getPixelFontName(), Font.PLAIN, 32);

    // Types
    @Getter
    private final AbstractGameType gameType;
    @Getter
    private final Environment environment;

    // Threads.
    private Thread autoSaveThread;
    private Thread collisionThread;
    private Thread ambientAudioThread;
    private Thread gameEventHandlerThread;

    // Files / folders.
    private File saveDir;

    // Active messages.
    private final ArrayList<String> activeMessages = new ArrayList<>();
    private final ArrayList<Long> activeMsgTimes = new ArrayList<>();

    // Current screen.
    @Getter
    @Deprecated
    private Screen currentScreen;

    // Audio.
    private AudioSlot ambientAudio;
    private long nextAudio;

    // Flags.
    @SuppressWarnings("FieldCanBeLocal")
    private boolean gameActive = false;

    // Event
    private EventBus.Handler binding;

    // Save
    @Getter
    private final SavedGame savedGame;
    private GameEvent currentGameEvent;

    public LoadedGame(SavedGame savedGame, Environment environment) {
        this.savedGame = savedGame;
        this.gameType = environment.getGameType();
        this.environment = environment;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Show and Hide     //
    ///////////////////////////
    public void run() {
        Util.setCursor(BubbleBlaster.getInstance().getBlankCursor());

        bindEvents();

        this.environment.getGameType().start();

        this.autoSaveThread = new Thread(this::autoSaveThread, "AutoSaver");
        this.autoSaveThread.start();

        this.collisionThread = new Thread(this::collisionThread, "Collision");
        this.collisionThread.start();

        this.ambientAudioThread = new Thread(this::ambientAudioThread, "Collision");
        this.ambientAudioThread.start();
    }

    @SuppressWarnings("deprecation")
    public void quit() {
        this.environment.quit();

        try {
            // Unbind events.
            gameType.unbindEvents();
            unbindEvents();

            autoSaveThread.interrupt();
            collisionThread.interrupt();
            ambientAudioThread.interrupt();
            gameEventHandlerThread.interrupt();

            environment.quit();

            // Hide cursor.
            Util.setCursor(BubbleBlaster.getInstance().getDefaultCursor());
        } catch (Exception e) {
            e.printStackTrace();
        }

        System.gc();
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Pause methods     //
    ///////////////////////////
    @Deprecated
    public void pause() {

    }

    @Deprecated
    public void unpause() {

    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Thread methods     //
    ////////////////////////////
    private void ambientAudioThread() {
//        Toolkit toolkit = Toolkit.getToolkit();

        while (this.gameActive) {
            if (this.ambientAudio == null) {
                if (!this.gameType.isBloodMoonActive() && this.nextAudio < System.currentTimeMillis()) {
                    if (new PseudoRandom(System.nanoTime()).getNumber(0, 5, -1) == 0) {
                        try {
                            this.ambientAudio = new AudioSlot(Objects.requireNonNull(getClass().getResource("/assets/qbubbles/audio/bgm/art_of_silence.mp3")), "artOfSilence");
                            this.ambientAudio.setVolume(0.25d);
                            this.ambientAudio.play();
                        } catch (URISyntaxException e) {
                            e.printStackTrace();
                            BubbleBlaster.getInstance().shutdown();
                        }
                    } else {
                        this.nextAudio = System.currentTimeMillis() + 1000;
                    }
                } else if (this.gameType.isBloodMoonActive()) {
                    try {
                        this.ambientAudio = new AudioSlot(Objects.requireNonNull(getClass().getResource("/assets/qbubbles/audio/bgm/blood_moon_state.mp3")), "bloodMoonState");
                        this.ambientAudio.setVolume(0.25d);
                        this.ambientAudio.play();
                    } catch (URISyntaxException e) {
                        e.printStackTrace();
                        BubbleBlaster.getInstance().shutdown();
                    }
                }
            } else if (this.ambientAudio.isStopped()) {
                this.ambientAudio = null;
            } else if (!this.ambientAudio.getClip().isPlaying()) {
                this.ambientAudio.stop();
                this.ambientAudio = null;
            } else if (this.gameType.isBloodMoonActive() && !this.ambientAudio.getName().equals("bloodMoonState")) {
                this.ambientAudio.stop();
                this.ambientAudio = null;

                try {
                    this.ambientAudio = new AudioSlot(Objects.requireNonNull(getClass().getResource("/assets/qbubbles/audio/bgm/blood_moon_state.mp3")), "bloodMoonState");
                    this.ambientAudio.setVolume(0.25d);
                    this.ambientAudio.play();
                } catch (URISyntaxException e) {
                    e.printStackTrace();
                    BubbleBlaster.getInstance().shutdown();
                }
            }
        }
    }

    private void collisionThread() {
        // Initiate variables for game loop.
        long lastTime = System.nanoTime();
        double amountOfUpdates = 30.0;
        double ns = 1000000000 / amountOfUpdates;
        @SuppressWarnings({"unused", "RedundantSuppression"}) double delta = 0;
        @SuppressWarnings({"unused", "RedundantSuppression"}) long timer = System.currentTimeMillis();
        @SuppressWarnings({"unused", "RedundantSuppression"}) int frames = 0;

        while (this.gameActive) {
            // Calculate tick delta-time.
            long now = System.nanoTime();
            delta += (now - lastTime) / ns;
            lastTime = now;

            List<Entity> entities = this.environment.getEntities();

            if (entities.isEmpty()) {
                System.out.println("ERROR");
            }

            List<Entity> loopingEntities = new ArrayList<>(entities);

            if (gameType.isInitialized()) {
                if (!BubbleBlaster.isPaused()) {
                    for (int i = 0; i < loopingEntities.size(); i++) {
                        for (int j = i + 1; j < loopingEntities.size(); j++) {
                            Entity source = loopingEntities.get(i);
                            Entity target = loopingEntities.get(j);
                            try {
                                // Check is collisionable with each other
                                if (source.isCollidingWith(target)) {

                                    // Check intersection.
                                    if (ShapeUtils.checkIntersection(source.getShape(), target.getShape())) {
                                        // Handling collision by posting collision event, and let the intersected entities attack each other.
                                        BubbleBlaster.getEventBus().post(new CollisionEvent(BubbleBlaster.getInstance(), delta, source, target));
                                        BubbleBlaster.getEventBus().post(new CollisionEvent(BubbleBlaster.getInstance(), delta, target, source));

                                        if (target instanceof DamageableEntity) {
                                            ((DamageableEntity) target).damage(source.getAttributeMap().getBase(Attribute.ATTACK) * delta / target.getAttributeMap().getBase(Attribute.DEFENSE), new DamageSource(source, DamageSourceType.COLLISION));
                                        }
                                        if (source instanceof DamageableEntity) {
                                            ((DamageableEntity) source).damage(target.getAttributeMap().getBase(Attribute.ATTACK) * delta / source.getAttributeMap().getBase(Attribute.DEFENSE), new DamageSource(target, DamageSourceType.COLLISION));
                                        }
                                    }
                                }
                            } catch (ArrayIndexOutOfBoundsException exception) {
                                BubbleBlaster.getLogger().info("Array index was out of bounds! Check check double check!");
                            }
                        }
                    }
                }
            }
            gameType.tick();
        }

        BubbleBlaster.getLogger().info("Collision thread will die now.");
    }

    private void autoSaveThread() {
        long nextSave = System.currentTimeMillis();
        while (!this.gameType.isGameOver()) {
//            System.out.println(nextSave - System.currentTimeMillis());
            if (nextSave - System.currentTimeMillis() < 0) {
                BubbleBlaster.getLogger().info("Auto Saving...");
                onAutoSave();
                nextSave = System.currentTimeMillis() + 30000;
            }
            try {
                //noinspection BusyWait
                Thread.sleep(30000);
            } catch (InterruptedException e) {
                BubbleBlaster.getLogger().warn("Could not sleep thread.");
            }
        }
        BubbleBlaster.getLogger().debug("Stopping AutoSaveThread...");
    }

    private synchronized void onAutoSave() {
        this.gameType.dumpSaveData(savedGame);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Event Binding     //
    ///////////////////////////
    public void bindEvents() {
        BubbleBlaster.getEventBus().register(this);
        this.gameActive = true;
    }

    public void unbindEvents() {
        this.binding.unbind();
        this.gameActive = false;
    }

    public boolean eventsAreActive() {
        return this.gameActive;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Getter-Only Properties     //
    ////////////////////////////////////
    public AudioSlot getAmbientAudio() {
        return this.ambientAudio;
    }

    public File getSaveDir() {
        return this.saveDir;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Trigger Methods     //
    /////////////////////////////
    public void triggerMessage(String s) {
        this.activeMessages.add(s);
        this.activeMsgTimes.add(System.currentTimeMillis() + 3000);
    }

    @Deprecated
    public synchronized void triggerGameOver() {
        if (!this.gameType.isGameOver()) {
            this.gameType.setResultScore(Math.round(Objects.requireNonNull(this.gameType.getPlayer()).getScore()));
            this.gameType.triggerGameOver();
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Render Methods     //
    ////////////////////////////

    /**
     * Render event
     *
     * @param game the game.
     * @param gg   the graphics instance.
     */
    @Deprecated
    public void renderGUI(BubbleBlaster game, GraphicsProcessor gg) {
        this.gameType.renderGUI(gg);

        if (this.currentScreen != null) this.currentScreen.renderGUI(game, new GraphicsProcessor(gg));
    }

    public void render(BubbleBlaster game, GraphicsProcessor gg) {
//        if (this.gameType.isBloodMoonActive()) {
//            gg.setColor(GameEvents.BLOOD_MOON_EVENT.get().getBackgroundColor());
//            gg.fillRect(0, 0, QBubbles.getInstance().getWidth(), QBubbles.getInstance().getHeight());
//        }
        if (this.gameType.isInitialized()) {
            this.gameType.render(gg);
        }
        this.renderHUD(game, gg);
    }

    public void renderHUD(@SuppressWarnings({"unused", "RedundantSuppression"}) BubbleBlaster game, GraphicsProcessor gg) {
        int i = 0;
        for (String s : activeMessages) {
            GraphicsProcessor gg2 = gg.create(0, 71 + (32 * i), 1000, 32);

            gg2.setColor(new Color(0, 0, 0, 128));
            gg2.fillRect(0, 0, 1000, 32);

            gg2.setColor(new Color(255, 255, 255, 255));
            GraphicsUtils.drawLeftAnchoredString(gg2, s, new Point2D.Double(2, 2), 28, defaultFont);

            gg2.dispose();
            i++;
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Events     //
    ////////////////////

    /**
     * <h1>Update Event-Handler</h1>
     * Update event handler for the game scene. Used for collision detection.
     */
    @SuppressWarnings("unchecked")
    @SubscribeEvent
    public void onUpdate() {
        @SuppressWarnings({"unused", "RedundantSuppression"}) ArrayList<String> messageCopy = (ArrayList<String>) this.activeMessages.clone();
        @SuppressWarnings({"unused", "RedundantSuppression"}) ArrayList<Long> timeCopy = (ArrayList<Long>) this.activeMsgTimes.clone();

        for (int i = 0; i < this.activeMessages.size(); i++) {
            if (this.activeMsgTimes.get(i) < System.currentTimeMillis()) {
                this.activeMsgTimes.remove(i);
                this.activeMessages.remove(i);
                i--;
            }
        }
    }

    /**
     * <h1>Keyboard Event-Handler</h1>
     * Keyboard event-handler, handles screen changing and debug-mode activations.
     *
     * @param evt the keyboard event.
     */
    @SubscribeEvent
    public void onKeyboard(KeyboardEvent evt) {
        if (evt.getType() == KeyEventType.PRESS) {
            if (evt.getParentEvent().getKeyCode() == KeyEvent.VK_ESCAPE) {
                if (!BubbleBlaster.isPaused()) {
                    BubbleBlaster.getInstance().displayScene(new PauseScreen());
                }
            } else if (evt.getParentEvent().getKeyCode() == KeyEvent.VK_SLASH) {
                if (!BubbleBlaster.isPaused()) {
                    BubbleBlaster.getInstance().displayScene(new CommandScreen());
                }
            } else if (evt.getParentEvent().getKeyCode() == KeyEvent.VK_F1 && BubbleBlaster.isDebugMode()) {
                this.gameType.triggerBloodMoon();
            } else if (evt.getParentEvent().getKeyCode() == KeyEvent.VK_F3 && BubbleBlaster.isDebugMode()) {
                Objects.requireNonNull(this.gameType.getPlayer()).destroy();
            } else if (evt.getParentEvent().getKeyCode() == KeyEvent.VK_F4 && BubbleBlaster.isDebugMode()) {
                Objects.requireNonNull(this.gameType.getPlayer()).levelUp();
            } else if (evt.getParentEvent().getKeyCode() == KeyEvent.VK_F5 && BubbleBlaster.isDebugMode()) {
                Objects.requireNonNull(this.gameType.getPlayer()).addScore(1000);
            } else if (evt.getParentEvent().getKeyCode() == KeyEvent.VK_F6 && BubbleBlaster.isDebugMode()) {
                Objects.requireNonNull(this.gameType.getPlayer()).setDamageValue(this.gameType.getPlayer().getMaxDamageValue());
            }
        } else if (evt.getType() == KeyEventType.TYPE) {
            if (evt.getKeyChar() == ' ') {
                if (this.gameType.isGameOver()) {
                    Util.getSceneManager().displayScreen(new GameOverScreen(gameType.getResultScore()));
                }
            }
        }
    }

    @SubscribeEvent
    public void onMouse(MouseEvent evt) {
        if (evt.getType() == MouseEventType.CLICK) {
            if (evt.getParentEvent().getButton() == 1) {
                if (KeyboardController.instance().isPressed(KeyEvent.VK_F1)) {
                    Objects.requireNonNull(this.gameType.getPlayer()).teleport(evt.getParentEvent().getPoint());
                }
            }
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Utility methods     //
    /////////////////////////////

    public void tick() {
        this.gameType.tick();
    }
}

File: src/main/kotlin/qtech/bubbles/LoadedGame.kt
package qtech.bubbles

import qtech.bubbles.common.Timer
import qtech.bubbles.common.Timer.time
import qtech.bubbles.common.entity.*
import qtech.bubbles.common.gamestate.GameEvent
import qtech.bubbles.common.gametype.AbstractGameMode
import qtech.bubbles.common.random.PseudoRandom
import qtech.bubbles.common.screen.Screen
import qtech.bubbles.core.common.SavedGame
import qtech.bubbles.core.controllers.KeyboardController
import qtech.bubbles.core.utils.categories.GraphicsUtils.drawLeftAnchoredString
import qtech.bubbles.core.utils.categories.ShapeUtils.checkIntersection
import qtech.bubbles.entity.DamageableEntity
import qtech.bubbles.entity.Entity
import qtech.bubbles.environment.Environment
import qtech.bubbles.event.KeyboardEvent
import qtech.bubbles.event.MouseEvent
import qtech.bubbles.event.SubscribeEvent
import qtech.bubbles.event.bus.EventBus
import qtech.bubbles.event.type.KeyEventType
import qtech.bubbles.event.type.MouseEventType
import qtech.hydro.GraphicsProcessor
import qtech.bubbles.media.AudioSlot
import qtech.bubbles.screen.CommandScreen
import qtech.bubbles.screen.GameOverScreen
import qtech.bubbles.screen.PauseScreen
import qtech.bubbles.util.Util.game
import qtech.bubbles.util.Util.sceneManager
import qtech.bubbles.util.Util.setCursor
import lombok.Getter
import java.awt.Color
import java.awt.Font
import java.awt.event.KeyEvent
import java.awt.geom.Point2D
import java.io.File
import java.net.URISyntaxException
import java.util.*

@Suppress("DEPRECATION", "UNUSED_PARAMETER", "UNUSED_VARIABLE", "UNCHECKED_CAST")
class LoadedGame(
// Save
    @field:Getter internal val savedGame: SavedGame, environment: Environment
) : Runnable {
    // Fonts.
    private val defaultFont = Font(game.pixelFontName, Font.PLAIN, 32)

    // Types
    @Getter
    val gameMode: AbstractGameMode

    @Getter
    private val environment: Environment

    // Threads.
    private var autoSaveThread: Thread? = null
    private var collisionThread: Thread? = null
    private var ambientAudioThread: Thread? = null
    private val gameEventHandlerThread: Thread? = null

    // Files / folders.
    val saveDir: File? = null

    // Active messages.
    private val activeMessages = ArrayList<String>()
    private val activeMsgTimes = ArrayList<Long>()

    // Current screen.
    @Getter
    @Deprecated("")
    private val currentScreen: Screen? = null

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Getter-Only Properties     //
    ////////////////////////////////////
    // Audio.
    var ambientAudio: AudioSlot? = null
        private set
    private var nextAudio: Long = 0

    // Flags.
    private var gameActive = false

    // Event
    private val binding: EventBus.Handler? = null
    private val currentGameEvent: GameEvent? = null

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Show and Hide     //
    ///////////////////////////
    override fun run() {
        setCursor(BubbleBlaster.instance.blankCursor)
        activate()
        environment.gameMode.start()
        autoSaveThread = Thread({ autoSaveThread() }, "AutoSaver")
        autoSaveThread!!.start()
        collisionThread = Thread({ collisionThread() }, "Collision")
        collisionThread!!.start()
        ambientAudioThread = Thread({ ambientAudioThread() }, "Collision")
        ambientAudioThread!!.start()
    }

    fun quit() {
        environment.quit()
        try {
            // Unbind events.
            gameMode.unbindEvents()
            deactivate()
            autoSaveThread!!.interrupt()
            collisionThread!!.interrupt()
            ambientAudioThread!!.interrupt()
            gameEventHandlerThread!!.interrupt()
            environment.quit()

            // Hide cursor.
            setCursor(BubbleBlaster.instance.defaultCursor)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        System.gc()
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Pause methods     //
    ///////////////////////////
    @Deprecated("")
    fun pause() {
    }

    @Deprecated("")
    fun unpause() {
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Thread methods     //
    ////////////////////////////
    private fun ambientAudioThread() {
//        Toolkit toolkit = Toolkit.getToolkit();
        while (gameActive) {
            if (ambientAudio == null) {
                if (!gameMode.isBloodMoonActive && nextAudio < System.currentTimeMillis()) {
                    if (PseudoRandom(System.nanoTime()).getNumber(0, 5, -1) == 0) {
                        try {
                            ambientAudio = AudioSlot(Objects.requireNonNull(javaClass.getResource("/assets/bubbleblaster/audio/bgm/art_of_silence.mp3")), "artOfSilence")
                            ambientAudio!!.volume = 0.25
                            ambientAudio!!.play()
                        } catch (e: URISyntaxException) {
                            e.printStackTrace()
                            BubbleBlaster.instance.shutdown()
                        }
                    } else {
                        nextAudio = System.currentTimeMillis() + 1000
                    }
                } else if (gameMode.isBloodMoonActive) {
                    try {
                        ambientAudio = AudioSlot(Objects.requireNonNull(javaClass.getResource("/assets/bubbleblaster/audio/bgm/blood_moon_state.mp3")), "bloodMoonState")
                        ambientAudio!!.volume = 0.25
                        ambientAudio!!.play()
                    } catch (e: URISyntaxException) {
                        e.printStackTrace()
                        BubbleBlaster.instance.shutdown()
                    }
                }
            } else if (ambientAudio!!.isStopped) {
                ambientAudio = null
            } else if (!ambientAudio!!.clip.isPlaying) {
                ambientAudio!!.stop()
                ambientAudio = null
            } else if (gameMode.isBloodMoonActive && ambientAudio!!.name != "bloodMoonState") {
                ambientAudio!!.stop()
                ambientAudio = null
                try {
                    ambientAudio = AudioSlot(Objects.requireNonNull(javaClass.getResource("/assets/bubbleblaster/audio/bgm/blood_moon_state.mp3")), "bloodMoonState")
                    ambientAudio!!.volume = 0.25
                    ambientAudio!!.play()
                } catch (e: URISyntaxException) {
                    e.printStackTrace()
                    BubbleBlaster.instance.shutdown()
                }
            }
        }
    }

    private fun collisionThread() {
        var time = time
        var passed: Double
        while (gameActive) {
            // Calculate tick delta-time.
            val time2 = Timer.time
            passed = time2 - time
            time = time2
            val entities: List<Entity> = environment.entities
            if (entities.isEmpty()) {
                println("ERROR")
            }
            val loopingEntities: List<Entity> = ArrayList(entities)
            if (gameMode.isInitialized) {
                if (!BubbleBlaster.isPaused) {
                    for (i in loopingEntities.indices) {
                        for (j in i + 1 until loopingEntities.size) {
                            val source = loopingEntities[i]
                            val target = loopingEntities[j]
                            try {
                                // Check is collisionable with each other
                                if (source.isCollidingWith(target)) {

                                    // Check intersection.
                                    if (checkIntersection(source.shape, target.shape)) {
                                        // Handling collision by posting collision event, and let the intersected entities attack each other.
//                                        BubbleBlaster.getEventBus().post(new CollisionEvent(BubbleBlaster.instance, passed, source, target));
//                                        BubbleBlaster.getEventBus().post(new CollisionEvent(BubbleBlaster.instance, passed, target, source));

//                                        LOGGER.debug("Attack / Defense: " + source.getAttributeMap().getBase(Attribute.ATTACK) + " / " + source.getAttributeMap().getBase(Attribute.DEFENSE));
//                                        LOGGER.debug("Attack / Defense: " + target.getAttributeMap().getBase(Attribute.ATTACK) + " / " + target.getAttributeMap().getBase(Attribute.DEFENSE));
                                        if (target is DamageableEntity) {
                                            if (source is DamageableEntity) {
                                                target.damage(source.attributeMap.get(Attribute.ATTACK) * passed / target.attributeMap.get(Attribute.DEFENSE), EntityDamageSource(source, DamageSourceType.COLLISION))
                                                source.damage(target.attributeMap.get(Attribute.ATTACK) * passed / source.attributeMap.get(Attribute.DEFENSE), EntityDamageSource(target, DamageSourceType.COLLISION))
                                            }
                                        }
                                    }
                                }
                            } catch (exception: ArrayIndexOutOfBoundsException) {
                                BubbleBlaster.logger.info("Array index was out of bounds! Check check double check!")
                            }
                        }
                    }
                }
            }
            gameMode.tick()
        }
        BubbleBlaster.logger.info("Collision thread will die now.")
    }

    private fun autoSaveThread() {
        var nextSave = System.currentTimeMillis()
        while (!gameMode.isGameOver) {
//            System.out.println(nextSave - System.currentTimeMillis());
            if (nextSave - System.currentTimeMillis() < 0) {
                BubbleBlaster.logger.info("Auto Saving...")
                onAutoSave()
                nextSave = System.currentTimeMillis() + 30000
            }
            try {
                Thread.sleep(30000)
            } catch (e: InterruptedException) {
                BubbleBlaster.logger.warn("Could not sleep thread.")
            }
        }
        BubbleBlaster.logger.debug("Stopping AutoSaveThread...")
    }

    @Synchronized
    private fun onAutoSave() {
        gameMode.dumpSaveData(savedGame)
        gameMode.dumpState(savedGame)
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Event Binding     //
    ///////////////////////////
    internal fun activate() {
        BubbleBlaster.eventBus.register(this)
        gameActive = true
    }

    internal fun deactivate() {
        binding!!.unbind()
        gameActive = false
    }

    fun isGameActive(): Boolean {
        return gameActive
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Trigger Methods     //
    /////////////////////////////
    fun triggerMessage(s: String) {
        activeMessages.add(s)
        activeMsgTimes.add(System.currentTimeMillis() + 3000)
    }

    @Deprecated("")
    @Synchronized
    fun triggerGameOver() {
        if (!gameMode.isGameOver) {
            gameMode.resultScore = Math.round(Objects.requireNonNull(gameMode.player)!!.score)
            gameMode.triggerGameOver()
        }
    }
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Render Methods     //
    ////////////////////////////
    /**
     * Render event
     *
     * @param game the game.
     * @param gg   the graphics instance.
     */
    @Deprecated("")
    fun renderGUI(game: BubbleBlaster?, gg: GraphicsProcessor?) {
        gameMode.renderGUI(gg)
        if (currentScreen != null) currentScreen.renderGUI(game!!, GraphicsProcessor(gg!!))
    }

    fun render(game: BubbleBlaster?, gg: GraphicsProcessor) {
//        if (this.gameType.isBloodMoonActive()) {
//            gg.setColor(GameEvents.BLOOD_MOON_EVENT.get().getBackgroundColor());
//            gg.fillRect(0, 0, QBubbles.getInstance().getWidth(), QBubbles.getInstance().getHeight());
//        }
        if (gameMode.isInitialized) {
            gameMode.render(gg)
        }
        renderHUD(game, gg)
    }

    fun renderHUD(game: BubbleBlaster?, gg: GraphicsProcessor) {
        var i = 0
        for (s in activeMessages) {
            val gg2 = gg.create(0, 71 + 32 * i, 1000, 32)
            gg2.color(Color(0, 0, 0, 128))
            gg2.rectF(0, 0, 1000, 32)
            gg2.color(Color(255, 255, 255, 255))
            drawLeftAnchoredString(gg2, s, Point2D.Double(2.0, 2.0), 28.0, defaultFont)
            gg2.dispose()
            i++
        }
    }
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Events     //
    ////////////////////
    /**
     * <h1>Update Event-Handler</h1>
     * Update event handler for the game scene. Used for collision detection.
     */
    @SubscribeEvent
    fun onUpdate() {
        val messageCopy = activeMessages.clone() as ArrayList<String>
        val timeCopy = activeMsgTimes.clone() as ArrayList<Long>
        var i = 0
        while (i < activeMessages.size) {
            if (activeMsgTimes[i] < System.currentTimeMillis()) {
                activeMsgTimes.removeAt(i)
                activeMessages.removeAt(i)
                i--
            }
            i++
        }
    }

    /**
     * <h1>Keyboard Event-Handler</h1>
     * Keyboard event-handler, handles screen changing and debug-mode activations.
     *
     * @param evt the keyboard event.
     */
    @SubscribeEvent
    fun onKeyboard(evt: KeyboardEvent) {
        if (evt.type === KeyEventType.PRESS) {
            if (evt.parentEvent.keyCode == KeyEvent.VK_ESCAPE) {
                if (!BubbleBlaster.isPaused) {
                    BubbleBlaster.instance.displayScene(PauseScreen())
                }
            } else if (evt.parentEvent.keyCode == KeyEvent.VK_SLASH) {
                if (!BubbleBlaster.isPaused) {
                    BubbleBlaster.instance.displayScene(CommandScreen())
                }
            } else if (evt.parentEvent.keyCode == KeyEvent.VK_F1 && BubbleBlaster.debugMode) {
                gameMode.triggerBloodMoon()
            } else if (evt.parentEvent.keyCode == KeyEvent.VK_F3 && BubbleBlaster.debugMode) {
                Objects.requireNonNull(gameMode.player)!!.destroy()
            } else if (evt.parentEvent.keyCode == KeyEvent.VK_F4 && BubbleBlaster.debugMode) {
                Objects.requireNonNull(gameMode.player)!!.levelUp()
            } else if (evt.parentEvent.keyCode == KeyEvent.VK_F5 && BubbleBlaster.debugMode) {
                Objects.requireNonNull(gameMode.player)!!.addScore(1000.0)
            } else if (evt.parentEvent.keyCode == KeyEvent.VK_F6 && BubbleBlaster.debugMode) {
                Objects.requireNonNull(gameMode.player)!!.damageValue = gameMode.player!!.maxDamageValue
            }
        } else if (evt.type === KeyEventType.TYPE) {
            if (evt.keyChar == ' ') {
                if (gameMode.isGameOver) {
                    sceneManager.displayScreen(GameOverScreen(gameMode.resultScore))
                }
            }
        }
    }

    @SubscribeEvent
    fun onMouse(evt: MouseEvent) {
        if (evt.type === MouseEventType.CLICK) {
            if (evt.parentEvent.button == 1) {
                if (KeyboardController.instance().isPressed(KeyEvent.VK_F1)) {
                    Objects.requireNonNull(gameMode.player)!!.teleport(evt.parentEvent.point)
                }
            }
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    //     Utility methods     //
    /////////////////////////////
    fun tick() {
        gameMode.tick()
    }

    init {
        gameMode = environment.gameMode
        this.environment = environment
    }
}